helix-lsp/src/lib.rs

```diff
@@ -263,83 +307,140 @@ pub mod util {
+                        blank = ""
+                    )
+                });
+
+                let (replacement, tabstops) =
+                    snippet::render(&snippet, &newline_with_offset, include_placeholder);
+
+                Some((
+                    (replacement_start, replacement_end, Some(replacement.into())),
+                    (mapped_replacement_start, tabstops),
+                ))
+            },
+            |range, tabstops| new_selection.push((*range, tabstops)),
+        );
+
+        let changes = transaction.changes();
+        if changes.is_empty() {
+            return transaction;
+        }
 
-        // Create new selection based on the cursor tabstop from above
-        let mut cursor_tabstop_offsets_iter = cursor_tabstop_offsets.iter();
-        let selection = selection
-            .clone()
-            .map(transaction.changes())
-            .transform_iter(|range| {
-                cursor_tabstop_offsets_iter
-                    .next()
-                    .unwrap()
+        let mut mapped_selection = SmallVec::with_capacity(new_selection.len());
+        let mut mapped_primary_idx = 0;
+        for (i, (range, (tabstop_anchor, tabstops))) in new_selection.into_iter().enumerate() {
+            if i == primary_idx {
+                mapped_primary_idx = mapped_selection.len()
+            }
+
+            let range = range.map(changes);
+            let tabstops = tabstops.first().filter(|tabstops| !tabstops.is_empty());
+            let Some(tabstops) = tabstops else{
+                // no tabstop normal mapping
+                mapped_selection.push(range);
+                continue;
+            };
+            // expand the selection to cover the tabstop to retain the helix selection semantic
+            // the tabstop closest to the range simply replaces `head` while anchor remains in place
+            // the remainaing tabstops recive their own single-width cursor
+            if range.head < range.anchor {
+                let primary_tabstop = tabstop_anchor + tabstops[0].1;
+                debug_assert!(primary_tabstop <= range.anchor);
+                let range = Range::new(range.anchor, primary_tabstop);
+                mapped_selection.push(range);
+                let rem_tabstops = tabstops[1..]
                     .iter()
-                    .map(move |(from, to)| {
-                        Range::new(
-                            (range.anchor as i128 + *from) as usize,
-                            (range.anchor as i128 + *to) as usize,
-                        )
-                    })
-            });
+                    .map(|tabstop| Range::point(tabstop_anchor + tabstop.1));
+                mapped_selection.extend(rem_tabstops);
+            } else {
+                let last_idx = tabstops.len() - 1;
+                let primary_tabstop = tabstop_anchor + tabstops[last_idx].1;
+                debug_assert!(primary_tabstop >= range.anchor);
```

Comment:
I tried this PR with rust-analyzer and no matter the snippet it crashed here!
```
thread 'main' panicked at 'assertion failed: primary_tabstop >= range.anchor', helix-lsp/src/lib.rs:428:17
```

Reply:
Interesting, it's a debug assert so in a release build it wouldnt't crash but it may mess up the selection a little. Could you share what your selection looked like?

Reply:
Hope this helps.
```
2023-03-09T09:57:38.253 helix_term::ui::completion [DEBUG] snippet text: "_mm256_and_ps(${1:a}, ${2:b})$0"
2023-03-09T09:57:38.253 helix_lsp::util [DEBUG] cursor=64 replacement_start=60 replacement_end=64 Snippet {
    elements: [
        Text(
            "_mm256_and_ps(",
        ),
        Placeholder {
            tabstop: 1,
            value: [
                Text(
                    "a",
                ),
            ],
        },
        Text(
            ", ",
        ),
        Placeholder {
            tabstop: 2,
            value: [
                Text(
                    "b",
                ),
            ],
        },
        Text(
            ")",
        ),
        Tabstop {
            tabstop: 0,
        },
    ],
}
2023-03-09T09:57:38.254 helix_lsp::util [DEBUG] primary_idx=0 Transaction {
    changes: ChangeSet {
        changes: [
            Retain(
                60,
            ),
            Insert(
                "_mm256_and_ps(a, b)",
            ),
            Delete(
                4,
            ),
            Retain(
                151,
            ),
        ],
        len: 215,
        len_after: 230,
    },
    selection: None,
}
2023-03-09T09:57:38.254 helix_lsp::util [DEBUG] Range {
    anchor: 79,
    head: 80,
    old_visual_position: None,
}
```

Reply:
ah interesting did you edit in append mode but without any selection? That's an edgecase I might not have considered and would fit this.

Definitly does help 👍

Reply:
I'm definitely in Insert mode. And the range printed is the one just here. (the `debug!` is placed just before the `debug_assert!`).

Reply:
Hmm mapping ranges should never change their direction, curious what's going on here. Will see if I can reprosucez thanks for testing!

Reply:
This should help you reproduce the issue.
Steps for reproducing:

- `cargo new --lib "hx-test"` (just a new empty project)
- `cd "hx-test"`
- `[..]/target/debug/hx -vv src/lib.rs`
- go to 2nd line and remove "right"
- then type "m256"
- then arrow down and crash!

Reply:
fixed it, I didn't handle the case where the selection was fully incased inside the removed text. Thanks for the pointers